gravitee-am-resource/gravitee-am-resource-infobip/src/main/java/io/gravitee/am/resource/infobip/provider/InfobipResourceProvider.java

```diff
@@ -0,0 +1,120 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(InfobipResourceProvider.class);
+
+    TfaApi tfaApi;
+    public static final String APPROVED = "approved";
+
+    private String pinId;
+    private String to;
+
+    @Autowired
+    private InfobipResourceConfiguration configuration;
+
+    @Override
+    public Completable send(MFALink target) {
+        this.to = target.getTarget();
+
+        String baseUrl = this.configuration.getBaseUrl();
+        String apiKey = this.configuration.getApiKey();
+        String apiKeyPrefix = this.configuration.getApiKeyPrefix();
+        String messageId = this.configuration.getMessageId();
+        String applicationId = this.configuration.getApplicationId();
+
+        ApiClient apiClient = new ApiClient();
+        apiClient.setApiKeyPrefix(apiKeyPrefix);
+        apiClient.setApiKey(apiKey);
+        apiClient.setBasePath(baseUrl);
+
+        this.tfaApi = new TfaApi(apiClient);
+
+        return Completable.create((emitter) -> {
+            try{
+                TfaStartAuthenticationResponse sendCodeResponse = this.tfaApi.sendTfaPinCodeOverSms(true,
+                        new TfaStartAuthenticationRequest()
+                                .applicationId(applicationId)
+                                .messageId(messageId)
+                                .from("InfoSMS")
+                                .to(this.to)
+                );
+
+                boolean isSuccessful = sendCodeResponse.getSmsStatus().equals("MESSAGE_SENT");
+
+                if(!isSuccessful) {
+                    emitter.onError(new SendChallengeException("Message not sent"));
+                }
+
+                this.pinId = sendCodeResponse.getPinId();
+
+                LOGGER.debug("Infobip Verification code asked with ID '{}'", sendCodeResponse.getPinId());
+                emitter.onComplete();
+            } catch (com.infobip.ApiException e) {
+                this.LOGGER.error("Challenge emission fails", e);
+                emitter.onError(new SendChallengeException("Unable to send challenge"));
+            }
+        });
+    }
+
+    @Override
+    public Completable verify(MFAChallenge challenge) {
+        return Completable.create((emitter) -> {
+            String pin = challenge.getCode();
+            try {
+                TfaVerifyPinResponse verifyResponse = this.tfaApi.verifyTfaPhoneNumber(pinId,
+                        new TfaVerifyPinRequest().pin(pin)
+                );
+                boolean verified = verifyResponse.getVerified();
+
+                LOGGER.debug(
+                        "Infobip Verification code with ID '{}' verified with status '{}'",
+                        this.pinId,
+                        verified);
+                if (!verified) {
+                    emitter.onError(new InvalidCodeException("Challenger not verified"));
+                }
+
+                emitter.onComplete();
```

Comment:
Put this code in a "else" section, otherwise both emitter.onError and  emitter.onComplete(); will be called if there is an error.